Домашняя работа 7

TatyanaV24 opened the pull request:
Решенные задачи по теме 7

TatyanaV24 committed:
Домашняя работа 7

Решенные задачи по теме 7

ios-review commented on HW7/HW7/BlueViewController.swift:
```diff
@@ -0,0 +1,29 @@
+    }
+    */
+
+}
```
Пустые контроллеры не нужны.
В сториборде можно оставить просто UIViewController и все будет хорошо работать.

ios-review commented on HW7/HW7/ItemThreeViewController.swift:
```diff
@@ -0,0 +1,38 @@
+    }
+    
+    @IBAction func colorGreenButton() {
+        (children[0] as? EmbedViewController)?.view.backgroundColor = UIColor.green
```
Обращаться по индексу не очень безопасно, так как такого элемента может не быть и приложение упадет.

Какие видите способы реализовать безопасно?

TatyanaV24 replied:
Вот тут я другого варианта пока не могу найти. При решении нашла только такой вариант, буду признательна за подсказку.

ios-review replied:
В функции prepare(for segue: UIStoryboardSegue, sender: Any?) есть объект этого контроллера.
Его можно сохранить в свойство, а затем использовать здесь.
В этой строке получаете инстанс встраиваемого контроллера.
`if let vc = segue.destination as? EmbedViewController`

ios-review commented on HW7/HW7/ItemThreeViewController.swift:
```diff
@@ -0,0 +1,38 @@
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        
+        if let vc = segue.destination as? EmbedViewController, segue.identifier == "embedShow"{
```
code style: нет пробела между "embedShow" и {

ios-review replied:
?

ios-review commented on HW7/HW7/ItemThreeViewController.swift:
```diff
@@ -0,0 +1,38 @@
+        
+        if let vc = segue.destination as? EmbedViewController, segue.identifier == "embedShow"{
+            vc.delegate = self
+            
```
code style: не нужный new line

ios-review commented on HW7/HW7/NextTwoViewController.swift:
```diff
@@ -0,0 +1,41 @@
+        }
+    
+    @IBAction func greenButton() {
+        self.view.window!.rootViewController?.dismiss(animated: true, completion: nil)
```
просто:
`dismiss(animated: true)`

ios-review commented on HW7/HW7/NextTwoViewController.swift:
```diff
@@ -0,0 +1,41 @@
+    @IBAction func redButton() {
+        self.view.window!.rootViewController?.dismiss(animated: true, completion: nil)
+        delegate?.changeText(_text: "Выбран \"красный цвет\"")
+    }
```
Предлагаю попробовать реализовать задачи с помощью модели данных, которую использовать во всех контроллерах приложения.

Для всего приложения создать модель ColorPallete.

```
enum ColorPallete: Int {
    case red = 0
    case green
    // etc

    var rgb: UIColor {
        // return appropriate value
    }

    var selectedDescription: String {
        // return appropriate value
    }

    var name: String {
        // return appropriate value
    }
}
```
У каждой кнопки задать tag, значение которого соответствует нужному кейсу ColorPallete.
У любой вью есть свойство tag, с помощью которого можно идентифицировать эту вью.
https://developer.apple.com/documentation/uikit/uiview/1622493-tag
Задать свойство можно как в коде, так и в сториборде.
Задайте соответствие tag кнопки и raw value для перечисления ColorPallete.
Создавать с помощью rawValue инициализатора
`let color = ColorPallete(rawValue: sender.tag)`

Для любой @IBAction функции можно назначить сколь угодно много кнопок для обработки нажатия.
Количество кода можно уменьшить, если обмениваться с другими контроллерами (через делегат) объектом перечисления ColorPallete.
А внутри контроллеров уже использовать вычислимые свойства этой модели для получения нужного значения: цвета или текста вроде "Выбран зеленый".

Будет примерно так:
```
@IBAction func buttonPressed(_ sender: UIButton) {
    guard let color = ColorPallete(rawValue: sender.tag) else { return }
    delegate?.update(color)
    dismiss(animated: true) 
}
```

Данную @IBAction функцию назначить для всех кнопок на контроллере.
Так сделать для всех контроллеров приложения.

ios-review reviewed (changes requested):
Работа с гит идеальна 🔥🔥🔥

Код полностью решает все поставленные задачи👍

Попробуете сделать его чуточку короче, реализовав модель данных?

TatyanaV24 committed:
Update BlueViewController.swift

Убрала пустой код, который не нужен. В прошлый раз пропустила при проверке.

TatyanaV24 committed:
Убрала пусту строку

TatyanaV24 committed:
Изменения в ItemTwoViewController

1. Добавила пробел после "showNext" и перед "{"
2. Внесла изменения в функцию changeText и  changeLabel.text

TatyanaV24 committed:
Модель enum

Добавила предложенную модель  enum, пока только в контроллер NextTwoViewController.
Задание не довела до конца, т.к. не смогла придумать, как именно использовать свойства в var rgb и т.д. В целом понимаю что такое свойства и методы перечислений, но в данном контексте не смогла реализовать полноценно. Нужна подсказка с Вашей стороны.

TatyanaV24 committed:
.

ios-review commented on HW7/HW7/BlueViewController.swift:
```diff
@@ -9,21 +9,6 @@ import UIKit
 
 class BlueViewController: UIViewController {
```
Класс контроллера все еще существует.
Он не нужен, если нет никакой логики.
В сториборде в поле Class можно использовать просто UIViewController и там настраивать параметры view.

ios-review replied:
<img width="1405" alt="Screenshot 2020-11-11 at 17 22 08" src="https://user-images.githubusercontent.com/62285448/98800114-87387e00-2442-11eb-8fef-374dd9ac9ac0.png">

ios-review commented on HW7/HW7/NextTwoViewController.swift:
```diff
@@ -7,11 +7,30 @@
+
+    var name: String {
+        return self.selectedDescription
+    }
```
code-style: отсупы

ios-review commented on HW7/HW7/NextTwoViewController.swift:
```diff
@@ -7,11 +7,30 @@
+
+    var selectedDescription: String { //выбранное описание
+        return 
+    }
```
данный код компилируется с ошибкой

ios-review commented on HW7/HW7/NextTwoViewController.swift:
```diff
@@ -7,11 +7,30 @@
+   var rgb: UIColor { // красный, зеленый, синий
+        
+    return UIColor.red; UIColor.green; UIColor.blue
+    }
```
Эти свойства нужны, чтобы преобразовывать конкретное значение ColorPallete в значение цвета или текста выбранного цвета.
```
    var rgb: UIColor { // красный, зеленый, синий
        switch self {
        case .red:
            return .red
        case .green:
            return .green
        case .blue:
            return .blue
        }
    }
```

ios-review commented on HW7/HW7/ItemTwoViewController.swift:
```diff
@@ -17,14 +17,14 @@ class ItemTwoViewController: UIViewController {
-    func changeText(_text:String){
-        changeLabel.text = "\(_text)"
+    func changeText(_text: ColorPallete){
+        changeLabel.text = "Выбран \(_text) цвет"
```
Например, здесь необходимо использовать свойство selectedDescription у ColorPallete.

ios-review replied:
Для этого необходимо реализовать это свойство.

ios-review replied:
`changeLabel.text = colorPallete.selectedDescription`

ios-review reviewed (changes requested):
В остальных контроллерах также нужно использовать ColorPallete + одну реализацию @IBAction функции.
В ней создавать ColorPallete по tag кнопки.
А затем использовать свойства ColorPallete для преобразования в текст или UIColor.

TatyanaV24 committed:
Добавление модели

Я наконец-то поняла, что именно требовалась. Первоначально, мне казалось, что необходимо использовать все три свойства в каждой задаче и они должны быть взаимосвязаны. И из-за этого я ходила по кругу. Но сейчас как мне кажется я направленом пути.

TatyanaV24 committed:
Модель и ошибки

Тут я попыталась также использовать модель, и на мой взгляд все выполнено верно. Тем не менее задача не срабатывает так , как я его этого хочу. Идея было в том, чтобы уменьшить количество контроллеров и оставить один. При нажатии на кнопку должен осуществляться переход на контроллер, который окрашивается в цвет соответствующий кнопке. Переход работает, а вот backgroungColor почему-то не изменяется.
Также дополнительно есть вопрос по модели. Допустим если ее применять к третьей задаче, то добавляется два новых цвета и соответсвенно XCode запросит прописать условия в других задачах и для этих цветом. Есть ли возможность обойти это?

ios-review commented on HW7/HW7/BlueViewController.swift:
```diff
@@ -7,8 +7,3 @@
-
-
-
-}
```
Пустые файлы также не нужны)
Ведь так?

ios-review commented on HW7/HW7/EmbedViewController.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    @IBAction func purpleButton() {
+        self.delegate?.didPickColor(color: .purple)
+    }
```
Создать только 1 @IBAction функцию с параметром sender.
Назначить для кнопок tag.
Назначить в качестве обработчика события нажатия для всех кнопок одну функцию.
В реализации создавать ColorPallete по tag.
Преобразовывать конкретное значение ColorPallete в UIColor с помощью свойства rgb.

ios-review commented on HW7/HW7/NextTwoViewController.swift:
```diff
@@ -0,0 +1,33 @@
+
+
+
+protocol NextDelegate {
```
code-style: 3 пустых строки, нужно не более одной.

ios-review commented on HW7/HW7/RedViewController.swift:
```diff
@@ -0,0 +1,9 @@
+//
+
+import UIKit
+
```
Пустой файл.

ios-review commented on HW7/HW7/ViewController.swift:
```diff
@@ -0,0 +1,54 @@
+
+import UIKit
+
+enum ColorPallete: Int {
```
Для данной модели создать отдельный файл с именем ColorPallete и перенести в этот файл реализацию ColorPallete.

ios-review reviewed (changes requested):
Модель данных описывает все данные, с которыми работает приложение.
Но не все свойства обязательно должны использоваться в каждом контроллере.
Для контроллера могут использоваться только определенные свойства модели данных.

Попробуете применять рекомендации сразу ко всем аналогичным местам в коде сразу?

TatyanaV24 committed:
Отдельный файл для модели ColorPallete

TatyanaV24 committed:
Изменения в EmbedViewController

Пересмотрела методы, привела все кнопки  каждого из контроллеров к одному методу.

TatyanaV24 committed:
Изменения в ItemThreeViewController

Также все кнопки свела к реализации через один метод.

TatyanaV24 committed:
Некорректная работа

Убрала лишние контроллеры, тем не менее реализация не проходит должным образом. Происходит переход, но нужный увет почему-то не отображается. Пыталась провести изменения экрана с использованием цикла, результат аналогичный. Это может быть связано с особенностью работы с Navigation Controller?

ios-review commented on HW7/HW7/ItemThreeViewController.swift:
```diff
@@ -9,29 +9,26 @@ import UIKit
+            self.subView?.view.backgroundColor = UIColor.purple
+        default:
+            break
+        }
```
Назначить кнопкам tag равные 1, 3 и 4.
Затем преобразовать tag в ColorPallete.
```
guard let color = ColorPallete(rawValue: sender.tag) else { return }
self.subView?.view.backgroundColor = color.rgb
```

ios-review commented on HW7/HW7/ItemThreeViewController.swift:
```diff
@@ -9,29 +9,26 @@ import UIKit
-    func didPickColor(color: UIColor){
-        self.dismiss(animated: true, completion: nil)
-        self.view.backgroundColor = color
+    @IBAction func transitionButton(_ sender: UIButton) {
```
Новую функцию необходимо переназначить в сториборде.
Сейчас приложение падает.

ios-review commented on HW7/HW7/ColorPalleteViewController.swift:
```diff
@@ -0,0 +1,46 @@
+//
+//  ColorPalleteViewController.swift
```
Просто ColorPallete.

ios-review commented on HW7/HW7/BlueViewController.swift:
```diff
@@ -1,9 +0,0 @@
-//
-
-import UIKit
-
```
Рекомендую добавлять/изменять/удалять файлы проекта только с помощью Xcode.
Если удалить с помощью файловой системы, то изменения не будут отображены в файле проекта .xcodeproj, который содержит метаданные о проекте, в том числе исполняемые файлы.

Удалив с помощью файловое системы, файл проекта останется неизменным и будет считать, что файл до сих пор есть в проекте, хотя по факту его не будет.
Удалив с помощью Xcode, файл проекта также будет изменен.
Удалить с помощью Xcode можно, выделив файл, нажать на del.

TatyanaV24 replied:
Удаление проводила именно по такой схеме, как Вы описали, возможно на git не все выгрузила..

ios-review commented on HW7/HW7/EmbedViewController.swift:
```diff
@@ -7,23 +7,11 @@
+    @IBAction func changeBackground(_ sender: UIButton) {
+        guard let color = ColorPallete(rawValue: sender.tag) else { return }
+        self.supView?.view.backgroundColor = color.rgb
     }
```
Новую функцию необходимо переназначить в сториборде.
Сейчас приложение падает.

ios-review reviewed (changes requested):
Задание было выполнено еще давно в соответствии с условиями!

Люблю добавлять свои "хотелки")
На скилбоксе задание приму.
Будет здорово, если все же доработаете его до конца.

TatyanaV24 committed:
Заключительные изменения

Провела изменения с учетом комментариев в git.
EmbedViewController и ItemThreeViewController - переназначила функции в сториборде. Странно, но перед отправкой проверяла приложение и все работало.
В ItemThreeViewController - добавила кнопкам tag hfdysq 1, 3 и 4 и исправила функцию, добавила преобразование tag в ColorPallete.

TatyanaV24 committed:
/

TatyanaV24 committed:
.

TatyanaV24 committed:
.

TatyanaV24 committed:
Finally

TatyanaV24 committed:
Merge branch 'HW7' of https://github.com/TatyanaV24/Swift-homework into HW7

TatyanaV24 committed:
Super finally
